### DISCUSSION TITLE
Retrieval of Events from Tenant Level Query

### DISCUSSION THREAD

**Discussion by dadams39** [microsoftgraph/msgraph-sdk-go#173] [Q&A] [answered]
## Retrieve Calendar Events from Tenant Level

Following along with [snippet](https://docs.microsoft.com/en-us/graph/api/calendar-list-calendarview?view=graph-rest-1.0&tabs=go)
It is noted that listing events on [these instructions]() indicate that events cannot be retrieved with application-level permissions. However, the first snippet says otherwise.  Assuming the first snippet is the most current. 

I have attempted to receive the events calendar events in the following configurations none have worked:
```golang
// Configuration 1:      startTime := "2022-01-01T19:00:00-08:00”  endTime := "2023-01-01T19:00:00-08:00"
result, err1 := client.UsersById(*userId).Calendar().CalendarView().GetWithRequestConfigurationAndResponseHandler(options, nil)
// Configuration 2
result, err1 := client.UsersById(*userId).Events().Get()
// Configuration 3
result, err1 := client.UsersById(*userId).Calendar().Get()
calendar := result.(models.Calendarable)
events := calendar.GetCalendarView()
```

In all cases the resulting `[]Eventable` is empty. If it is possible to retrieve calendar events from the tenant level, could you advise on the changes that need to be made?

**Comment by baywet** [comment]
are you getting any error?

**Reply by dadams39** [reply]
For configuration 1 and 2:
```
fmt.Printf("Type: %T\n", result)
fmt.Printf("Err?: %v\n", err1)
```
Returns:
```
Type: <nil>
Err?: Unknown LocationType value: default
```
Configuration 3 returns calenderable instance. The call to events returns `nil`

**Reply by baywet** [reply]
is it possible the inbox you're targeting has no events setup?
in scenario 3, are you able to get properties of the calendar itself like its name? (by default the events are not going to be returned, you need to use the expand query parameter and add "events" in there).

**Reply by dadams39** [reply]
I am able to get the name and other aspects of the calendar and print them. 
<img width="643" alt="Screen Shot 2022-05-24 at 10 01 24 AM" src="https://user-images.githubusercontent.com/8600618/170113151-0852131e-4a94-4a85-bc87-3b9f2f58e1e7.png">
I created several repeating meetings and am not sure of your meaning for expanding the query.

If I were using delegated permissions, I would use the snippet from Graph Explorer:
```go
requestParameters := &msgraphsdk.EventsRequestBuilderGetQueryParameters{
	Select: "subject,body,bodyPreview,organizer,attendees,start,end,location",
}
options := &msgraphsdk.EventsRequestBuilderGetRequestConfiguration{
	QueryParameters: requestParameters,
}
result, err := graphClient.Me().Events().GetWithRequestConfigurationAndResponseHandler(options, nil)
```
Is the suggestion to use the Events request or to somehow expand the calendar request to include events?

**Reply by baywet** [reply]
nevermind this suggestion, this endpoint doesn't seem to support expand...
You won't be able to get events through just calendar, you need to write `client.UsersById(*userId).Calendar().Events().Get()` instead.

If you run the following code, what do you get?

```golang
request := client.UsersById(*userId).Events().CreateGetRequestInformation()
request.PathParameters["baseurl"] = "https://graph.microsoft.com/v1.0"
url, err := request.GetUri()
```

for each of the scenarios (I want to check the URLs are built properly).

Also if you try to switch your application and credentials for a delegated context (let's say device code flow), do you get results?

**Reply by dadams39** [reply]
From the tenant level. The response:
```
url: https://graph.microsoft.com/v1.0/users/f435c656-f8b2-4d71-93c3-6e092f52a167/events
err: <nil>
```
From the delegated user level: **Interactive Credential** 
```
url: https://graph.microsoft.com/v1.0/users/f435c656-f8b2-4d71-93c3-6e092f52a167/events
err: <nil>
```
I also want to affirm that the delegated user does NOT receive the events either.
The OdataError is `Unknown LocationType value: default`

**Reply by baywet** [reply]
The urls seem correct. I'm not sure why you're not getting any information. If you try the same request with Microsoft graph explorer, what do you get?

**Reply by dadams39** [reply]
The code snippet from  Graph is: 
`https://graph.microsoft.com/v1.0/me/calendar/events?` <--> `result, err := graphClient.Me().Calendar().Events().Get()`
which returns all of the events. 
Should this be made into an issue? 

Filling in the complete information also returns all of the events:
`https://graph.microsoft.com/v1.0/users/{user-id}/calendar/events` --> 

`result, err := graphClient.UsersById(&userId).Calendar().Events().Get()`
Which also returns all the associated Events.

**Reply by baywet** [reply]
when you say "which returns all the events" do you mean only in graph explorer or also for the Go equivalents?

**Reply by dadams39** [reply]
I am saying the Graph Explorer input returns all the events in both cases. The derivatives in Golang are not working. Has anyone else verified my experience?

**Reply by baywet** [reply]
I just have, the following snippet

```golang
cred, err := azidentity.NewDeviceCodeCredential(&azidentity.DeviceCodeCredentialOptions{
		TenantID: "<UPDATE THE TENANT ID>",
		ClientID: "<UPDATE THE APP ID>",
		UserPrompt: func(ctx context.Context, message azidentity.DeviceCodeMessage) error {
			fmt.Println(message.Message)
			return nil
		},
	})
	if err != nil {
		fmt.Printf("Error creating credentials: %v\n", err)
	}
	auth, err := a.NewAzureIdentityAuthenticationProviderWithScopes(cred, []string{"Calendars.Read"})
	if err != nil {
		fmt.Printf("Error authentication provider: %v\n", err)
		return
	}
	adapter, err := sdk.NewGraphRequestAdapter(auth)
	if err != nil {
		fmt.Printf("Error creating adapter: %v\n", err)
		return
	}
	client := sdk.NewGraphServiceClient(adapter)

	result, err := client.Me().Calendar().Events().Get()

	if err != nil {
		fmt.Printf("Error executing request: %v\n", err)
		return
	}
	fmt.Printf("result = %+v\n", *(result.GetValue()[0].GetSubject()))
	
```

returns 

```
To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code REPLACED to authenticate.
result = test event tomorrow
```

With an account which has an event for the next day

**Reply by dadams39** [reply]
My original question was for signing in at the application level or the tenant.... the CalendarView stated it was retrievable from the application level with application permissions instead of delegated. This appears that is untrue.

**Reply by baywet** [reply]
I understand the authentication aspect of the question. We've validated the sdk builds urls properly and deserializes the response properly. It becomes either an authentication issue, in which the which case you should reach out to https://github.com/azure/azure-sdk-for-go.
Or it's a documentation issue, and you can create an issue from the documentation page. Or its an Api behaviour issue, in which case you should reach out on the dedicated Microsoft Q&A forum to get attention from the service team.

**Comment by dadams39** [comment, marked as answer]
Here is the full download of this problem of the issue. **For the Tl;dr** Use a selector until Get is fixed
When making this call from the tenant level:
`result, err1 := client.UsersById(*userId).Events().Get()`
I describe the print statements and below show the output
![Screen Shot 2022-05-26 at 2 02 45 PM](https://user-images.githubusercontent.com/8600618/170548965-432c0de7-2eaa-4863-85dd-dd99f9657a64.png)
If you are confused, stay with me. It gets better. So, in  [models/events.go](https://github.com/microsoftgraph/msgraph-sdk-go/blob/main/models/event.go), there are fields `location` and `locations`. When the Get() function tries to return `location`, you get this error. So, use `msevt.EventsRequestBuilderGetRequestConfiguration` 
where -->  `msevt "github.com/microsoftgraph/msgraph-sdk-go/users/item/events"` and do not select the location. 

And you will be able to get the event list. I cannot claim to know why you get this behavior when using application permissions, but I posted the full code snippet below. @baywet I do believe this is actually a bug.

```
func GetEventOptions() *msevt.EventsRequestBuilderGetRequestConfiguration {
        selecting := []string{"id,allowNewTimeProposals,attendees,body,bodyPreview,calendar,end,extensions,hasAttachments,hideAttendees,iCalUId,importance,instances,isAllDay,isCancelled,isDraft,isOnlineMeeting,isOrganizer,isReminderOn,locations,onlineMeeting,onlineMeetingProvider,onlineMeetingUrl,organizer,originalEndTimeZone,originalStart,originalStartTimeZone,recurrence,reminderMinutesBeforeStart,responseRequested,responseStatus,sensitivity,seriesMasterId,showAs,singleValueExtendedProperties,start,subject,transactionId,webLink"}
        requestParameters := &msevt.EventsRequestBuilderGetQueryParameters{
                Select: selecting,
        }
        options := &msevt.EventsRequestBuilderGetRequestConfiguration{
                QueryParameters: requestParameters,
        }
        return options
}
```

**Reply by baywet** [reply]
aaaah very good catch, the error message actually comes from [here](https://github.com/microsoftgraph/msgraph-sdk-go/blob/5ab44d02be0598d49f95a73464dd57f06da1817b/models/location_type.go#L48) my guess is that you're still on version 0.18.0 which [contains the bug](https://github.com/microsoftgraph/msgraph-sdk-go/blob/0bb5af57a5d7c910cafe5cf5ebac8aa6337b07aa/models/location_type.go#L28)

So long story short, this has nothing to do with the authentication scenario but it's tied to the tenant data (my test data didn't happen to have a location) and you should upgrade to the latest version of the SDK (0.24.1 at the time of writing)
